### PR TITLE
Add InputDetect Lib and patches

### PR DESCRIPTION
**Pull request by ReveredRedHood**
Tested in-game and GetButtonDown worked. GetButtonUp didn't, because the game doesn't seem to call Rewired.Player.GetButtonUp(...), so a more creative solution may be needed for detecting when a button is released instead of pressed.

### CODE REVIEW

File: SOD.Common/Patches/RewiredPlayerPatches.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            internal static int lastFrameCount = -1;
+
+            [HarmonyTargetMethod]
+            internal static MethodBase CalculateMethod()
+            {
+                return typeof(Rewired.InputManager_Base).Assembly.ExportedTypes
+                    .Single(t => t.FullName == "Rewired.Player")
+                    .GetMethods()
+                    .Where(m => m.Name == "GetButtonDown")
+                    .Single(m => m.GetParameters().Single().ParameterType == typeof(string));
```
Comment (Ven0maus) [inline SOD.Common/Patches/RewiredPlayerPatches.cs:25]: rework to FirstOrDefault() with descriptive error messages, (I would not use Single or SingleOrDefault) as they will still give sequence errors if there is more than 1 element

File: SOD.Common/Patches/RewiredPlayerPatches.cs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            internal static int lastFrameCount = -1;
+
+            [HarmonyTargetMethod]
+            internal static MethodBase CalculateMethod()
+            {
+                return typeof(Rewired.InputManager_Base).Assembly.ExportedTypes
+                    .Single(t => t.FullName == "Rewired.Player")
+                    .GetMethods()
+                    .Where(m => m.Name == "GetButtonUp")
+                    .Single(m => m.GetParameters().Single().ParameterType == typeof(string));
```
Comment (Ven0maus) [inline SOD.Common/Patches/RewiredPlayerPatches.cs:57]: Same as above

File: SOD.Common/Shadows/Implementations/InputDetection.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    // Figure out how the game does shift key (sprint toggle) detection
+    // Potentially add mouse axis detection, etc.
+    // The game doesn't call Rewired.Player.GetButtonUp(...) for anything, so the OnButtonUp event is sort of pointless...
+    // but it would be good to find a way to hook into OnButtonUp eventually.
+    public sealed class InputDetection
+    {
+        /// <summary>
+        /// Raised once when a bound action changes from being released/up to being pressed/down.
+        /// </summary>
+        public event EventHandler<InputDetectionEventArgs> OnButtonDown;
```
Comment (Ven0maus) [inline SOD.Common/Shadows/Implementations/InputDetection.cs:15]: Change to OnButtonStateChanged

File: SOD.Common/Shadows/Implementations/InputDetection.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+        internal void OnButtonStateChanged(string actionName, bool isDown)
+        {
+            // if (isDown)
+            OnButtonDown?.Invoke(this, new InputDetectionEventArgs(actionName));
```
Comment (Ven0maus) [inline SOD.Common/Shadows/Implementations/InputDetection.cs:20]: Pass isDown to event args

**Comment by Ven0maus** [comment]
Looks good, left some comments as reminders to what was left to do

File: SOD.Common/Patches/InputPatches.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                        actionNames.Add(action.name);
+                    }
+                }
+                foreach (var actionName in actionNames)
+                {
+                    if (__instance.player.GetButtonDown(actionName))
+                    {
+                        Lib.InputDetection.ReportButtonStateChange(actionName, true);
+                    }
+                    if (__instance.player.GetButtonUp(actionName))
```
Comment (Ven0maus) [inline SOD.Common/Patches/InputPatches.cs:45]: Might be worth using elseif here, I don't think both can happen at the same time for the same action?

File: SOD.Common/Patches/RewiredPlayerPatches.cs (hunk 1 above)
Comment (Ven0maus) [reply SOD.Common/Patches/RewiredPlayerPatches.cs]: N/A

File: SOD.Common/Patches/RewiredPlayerPatches.cs (hunk 2 above)
Comment (Ven0maus) [reply SOD.Common/Patches/RewiredPlayerPatches.cs]: N/A

File: SOD.Common/Patches/RewiredPlayerPatches.cs (hunk 1 above)
Comment (ReveredRedHood) [reply SOD.Common/Patches/RewiredPlayerPatches.cs]: No longer needed with changes to patch

File: SOD.Common/Patches/RewiredPlayerPatches.cs (hunk 2 above)
Comment (ReveredRedHood) [reply SOD.Common/Patches/RewiredPlayerPatches.cs]: No longer needed with changes to patch

**Pull request merged**